src/bin/client-demo.rs

```diff
@@ -134,85 +187,84 @@ fn main() {
-            txs.len(),
-            leader.transactions_addr
-        );
-        let client = mk_client(&client_addr, &leader);
-        for tx in txs {
-            client.transfer_signed(tx.clone()).unwrap();
-        }
-    });
+    println!("Sampling tps every second...",);
 
+    let maxes = Arc::new(RwLock::new(Vec::new()));
     let sample_period = 1; // in seconds
-    println!("Sampling tps every second...",);
-    let maxes: Vec<_> = validators
-        .into_par_iter()
-        .map(|val| {
-            let mut client = mk_client(&client_addr, &val);
-            let mut now = Instant::now();
-            let mut initial_tx_count = client.transaction_count();
-            let mut max_tps = 0.0;
-            let mut total = 0;
-            for i in 0..100 {
-                let tx_count = client.transaction_count();
-                let duration = now.elapsed();
-                now = Instant::now();
-                let sample = tx_count - initial_tx_count;
-                initial_tx_count = tx_count;
-                println!(
-                    "{}: Transactions processed {}",
-                    val.transactions_addr, sample
-                );
-                let ns = duration.as_secs() * 1_000_000_000 + u64::from(duration.subsec_nanos());
-                let tps = (sample * 1_000_000_000) as f64 / ns as f64;
-                if tps > max_tps {
-                    max_tps = tps;
-                }
-                println!("{}: {} tps", val.transactions_addr, tps);
-                total = tx_count - first_count;
-                println!(
-                    "{}: Total Transactions processed {}",
-                    val.transactions_addr, total
-                );
-                if total == transactions.len() as u64 {
-                    break;
-                }
-                if i > 20 && sample == 0 {
-                    break;
-                }
-                sleep(Duration::new(sample_period, 0));
-            }
-            (max_tps, total)
+    let v_threads: Vec<_> = validators
+        .into_iter()
+        .map(|v| {
+            let exit = signal.clone();
+            let thread_addr = client_addr.clone();
+            let maxes = maxes.clone();
+            Builder::new()
+                .name("solana-client-sample".to_string())
+                .spawn(move || {
+                    sample_tx_count(thread_addr, exit, maxes, first_count, v, sample_period);
+                })
+                .unwrap()
         })
         .collect();
+
+    let time = Duration::new(time_sec, 0);
+    let now = Instant::now();
+    while now.elapsed() < time {
+        println!(
+            "Signing transactions... {} {}",
+            keypair_pairs.len(),
+            keypair_pairs[0].len()
+        );
+        let signing_start = Instant::now();
+        let transactions: Vec<_> = keypair_pairs
+            .par_iter()
+            .map(|chunk| Transaction::new(&chunk[0], chunk[1].pubkey(), 1, last_id))
```

Comment:
I'm curious, how are you calculating TPS if TX signing is inside the loop?

Reply:
the other thread is sampling it.